### PR TITLE
Add retries to Graphql client

### PR DESCRIPTION
**Pull request by aszecowka**
In this PR we are trying to address `connection refused` error.
According to [this discussion](https://serverfault.com/a/725263) , this error occurs in 2 cases:
- no process does listen on the given port
- a port is blocked by the firewall 

I was able to reproduce this problem by executing compass integration tests multiple times (with setting `count` to ClusterTestSuite). 
The communication problem is not caused by Director Pod, but rather by a testing pod. Testing pod has a sidecar injected, we have to wait until it is fully functional and we can make calls from it.
In Kyma, there is a pattern to run tests in the following way:

```
 - sleep 10; /director.test -test.v; exit_code=$?; pkill -INT pilot-agent;
   sleep 4; exit $exit_code;
```
As you can see, at the beginning there is sleep for 10 seconds, to give Istio time to make Pod fully functional. Unfortunately, sometimes 10 seconds is not enough.
In this PR, I added retries for GraphQL calls, when `connection refused` occurs.
I executed 100 times Compass tests:
```
Testing Pod 23
time="2019-10-16T07:25:30Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
time="2019-10-16T07:25:31Z" level=warning msg="[testContext] OnRetry: attempts: 1, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"

Testing Pod 46
time="2019-10-16T07:40:17Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
```
As you can see, retries were used 2 times and thanks to that all 100 test executions passed.
If I reduce sleep time to 5 seconds before starting test, retries are more frequent:
```
Pod 0
time="2019-10-16T08:16:29Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 1
time="2019-10-16T08:17:04Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
time="2019-10-16T08:17:05Z" level=warning msg="[testContext] OnRetry: attempts: 1, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 2
time="2019-10-16T08:17:42Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 3
time="2019-10-16T08:18:14Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 4
time="2019-10-16T08:18:52Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 5
Pod 6
Pod 7
Pod 8
time="2019-10-16T08:21:06Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 9
time="2019-10-16T08:21:40Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
time="2019-10-16T08:21:41Z" level=warning msg="[testContext] OnRetry: attempts: 1, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 10
Pod 11
Pod 12
time="2019-10-16T08:23:25Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 13
time="2019-10-16T08:24:03Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 14
Pod 15
time="2019-10-16T08:25:15Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 16
time="2019-10-16T08:25:48Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 17
time="2019-10-16T08:26:22Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
Pod 18
Pod 19
time="2019-10-16T08:27:26Z" level=warning msg="[testContext] OnRetry: attempts: 0, error: Post http://compass-director.compass-system.svc.cluster.local:3000/graphql: dial tcp 10.11.247.46:3000: connect: connection refused\n"
```

### CODE REVIEW

File: tests/end-to-end/README.md [hunk 1]
```diff
@@ -7,4 +7,15 @@ Director folder contains tests of GraphQL API for managing Applications and Runt
 To run Director tests, execute:
 ```
 ./run.sh
-```
\ No newline at end of file
+```
+
+## TODO environment variables
```
Comment (aszecowka) [inline tests/end-to-end/README.md]: TODO: remove TODO

File: tests/end-to-end/README.md [hunk 2]
```diff
@@ -7,4 +7,12 @@ Director folder contains tests of GraphQL API for managing Applications and Runt
...
\ No newline at end of file
+```
+
+## Configuration
+
+You can specify following environment variables.
+
+| ENV                                      | Default                         | Description                                               |
+| ---------------------------------------- | ------------------------------- | --------------------------------------------------------- |
+| DIRECTOR_URL                             | 127.0.0.1:3000/                  | The address and port for the Director                     |
```
Comment (pkosiec) [inline tests/end-to-end/README.md]: Default is `127.0.0.1:3000`, withouth slash

File: tests/end-to-end/director/testctx.go [hunk 3]
```diff
@@ -61,7 +66,19 @@ func (tc *testContext) RunOperation(ctx context.Context, req *gcli.Request, resp
...
+		return tc.cli.Run(ctx, req, &m)
+	})
+}
+
+func (tc *testContext) withRetryOnConnectionRefused(risky func() error) error {
+	return retry.Do(risky, retry.Attempts(7), retry.Delay(time.Second), retry.OnRetry(func(n uint, err error) {
+		logrus.WithField("component", "testContext").Warnf("OnRetry: attempts: %d, error: %v", n, err)
+
+	}), retry.LastErrorOnly(true), retry.RetryIf(func(err error) bool {
+		return strings.Contains(err.Error(), "connection refused")
```
Comment (pkosiec) [inline tests/end-to-end/director/testctx.go]: hmm, what about other cases than `connection refused`?

File: tests/end-to-end/gen-examples.sh [hunk 4]
```diff
@@ -121,7 +121,7 @@ rm -f "${ROOT_PATH}/examples/"*
 
 echo -e "${GREEN}Running Director tests with generating examples...${NC}"
```
Comment (pkosiec) [inline tests/end-to-end/gen-examples.sh]: Please use the `wait-for-director.sh` script in `gen-examples.sh`

File: tests/end-to-end/director/testctx.go (hunk 3 above)
Comment (pkosiec) [reply tests/end-to-end/director/testctx.go]: For example "connection reset by peer"

**Pull request merged**